geoopt/manifolds/symmetric_positive_definite.py

```diff
@@ -0,0 +1,385 @@
+from functools import partial
+from typing import Callable, Optional, Tuple, Union
+import torch
+from .base import Manifold
+
+__all__ = ["SymmetricPositiveDefinite"]
+
+
+EPS = {torch.float32: 1e-4, torch.float64: 1e-7}
+
+
+class SymmetricPositiveDefinite(Manifold):
+    r"""Manifold of symmetric positive definite matrices.
+
+    .. math::
+
+        A = A^T\\
+        \langle x, A x \rangle > 0 \quad , \forall x \in \mathrm{R}^{n}, x \neq 0 \\
+        A \in \mathrm{R}^{n\times m}
+
+
+    The tangent space of the manifold contains all symmetric matrices.
+
+    References
+    ----------
+    - https://github.com/pymanopt/pymanopt/blob/master/pymanopt/manifolds/psd.py
+    - https://github.com/dalab/matrix-manifolds/blob/master/graphembed/graphembed/manifolds/spd.py
+
+    Parameters
+    ----------
+    ndim : int
+        number of trailing dimensions treated as matrix dimensions.
+        All the operations acting on such as inner products, etc
+        will respect the :attr:`ndim`.
+    """
+
+    __scaling__ = Manifold.__scaling__.copy()
+    name = "SymmetricPositiveDefinite"
+    ndim = 0
+    reversible = True
+    defaulf_metric: str = "AIM"
+
+    def __init__(self, ndim=2, defaulf_metric: str = defaulf_metric):
+        super().__init__()
+        self.ndim = ndim
+        self.defaulf_metric = defaulf_metric
+
+    def _funcm(
+        self, x: torch.Tensor, func: Callable[[torch.Tensor], torch.Tensor]
+    ) -> torch.Tensor:
+        """Apply function to symmetric matrix.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+        func : Callable[[torch.Tensor], torch.Tensor]
+            function to apply
+
+        Returns
+        -------
+        torch.Tensor
+            symmetric matrix with function applied to
+        """
+        e, v = torch.symeig(x, eigenvectors=True)
+        return v @ torch.diag_embed(func(e)) @ v.transpose(-1, -2)
+
+    def _expm(self, x: torch.Tensor, using_native=False) -> torch.Tensor:
```

Comment:
what is the case when we have to choose? btw, I have an implemented [`expm`](https://github.com/geoopt/geoopt/blob/master/geoopt/linalg/batch_linalg.py#L50). This implementation was waiting for https://github.com/pytorch/pytorch/issues/9983 and seems like it is partially resolved in the new release. Feel free to conditionally replace it (right in `geoopt/linalg/batch_linalg.py`) based on availability of `torch.matrix_exp`.